content/300-guides/050-database/830-prismas-type-system.mdx

```diff
@@ -0,0 +1,154 @@
+---
+title: How to use Prisma's type system
+metaDescription: How to use Prisma's type system
+tocDepth: 2
+---
+
+<TopBlock>
+
+This guide explains Prisma's type system and walks you through using it to introspect your database and apply schema changes.
+
+</TopBlock>
+
+## How does Prisma's type system work?
+
+Prisma uses types to define the kind of data that a field can hold. To make it easy to get started, Prisma provides a small number of core [scalar types](/reference/api-reference/prisma-schema-reference#model-field-scalar-types). For example, in the blog post model below, the `title` field uses the `String` scalar type, while the `createdAt` field uses the `DateTime` scalar type:
+
+```prisma file=schema.prisma
+model Post {
+  id        Int      @id
+  title     String
+  createdAt DateTime
+}
+```
+
+Databases also have their own type system, which defines the type of value that a column can hold. Most databases provide a large number of data types to allow fine-grained control over exactly what a column can store. For example, a database might provide inbuilt support for multiple sizes of integers, or for XML data. The names of these types vary between databases. In PostgreSQL a variable length text column has a type of `text`, whereas in MySQL the type is `varchar`.
+
+### Default type mappings
+
+To allow you to get started with our core scalar types, Prisma provides _default type mappings_ that map each scalar type to a default type in the underlying database. For example:
+
+- by default Prisma's `String` type gets mapped to PostgreSQL's `text` type and MySQL's `varchar` type
+- by default Prisma's `DateTime` type gets mapped to PostgreSQL's `timestamp(3)` type and SQL Server's `datetime2` type
+
+See Prisma's [database connector pages](/concepts/database-connectors) for the default type mappings for a given database. For example, [this table](/concepts/database-connectors/postgresql#prisma-migrate) gives the default type mappings for PostgreSQL. To see the default type mappings for a given database, see the [model field scalar types section]() of the Prisma schema reference. For example, [this table](/reference/api-reference/prisma-schema-reference#model-field-scalar-types) gives the default type mappings for the `String` scalar type.
+
+### Native type mappings
+
+Sometimes you may need to use a more specific database type that is not one of the default type mappings for your Prisma type. For this purpose, Prisma provides [native type attributes](/concepts/components/prisma-schema/data-model#native-types-mapping) to refine the core scalar types. For example, in the `createdAt` field of your `Post` model above you may want to use a timestamp with a time zone in an underlying PostgreSQL database, by using the `timestamptz(3)` type instead of the default type mapping of `timestamp(3)`. To do this, add a `@db.Timestamptz(3)` native type attribute to the `createdAt` field:
```

Comment:
This is difficult to grok, as the database type `timestamptz(3)` and the Prisma native type `@db.Timestamptz(3)` looks slightly different, but both look pretty similar to the default PostgreSQL `timestamp(3)`. I bet some people would overlook the `tz` and be left wondering what is going on.

(I unfortunately do not have a solution handy for this problem. Maybe a native type that is more different from the default?)

Reply:
I also considered using something like `@db.Date` instead, to be more visibly different, though it seems more contrived as a type choice? 

I wish I'd written down my thought process when I was choosing an example, but I was having difficulty picking a good one. Everything I thought of seemed to have something different wrong with it:

- `@db.citext` doesn't require much explanation, but needs a Postgres extension
- `@db.Uuid` requires explaining what UUIDs are, plus reading our docs it felt kind of confusing because we also have the `uuid()` function
- anything to do with money runs in to a lot of opinions about how you should represent money
- some other things I forgot...

Happy to have other suggestions!

Reply:
Maybe strings could work? The default string native type could very easily be very different from some of the other options available.

Reply:
Strings are good from the point of the view of the names looking different:

<img width="719" alt="image" src="https://user-images.githubusercontent.com/3910386/191705714-e0c5f25b-ca21-467f-b4de-8a8405a89a3f.png">

I don't really know enough to pick a plausible example though - do you have any ideas out of those?

Reply:
A super specific case would be someone wanting to have `String` with long text (so `@db.Text`) and another `String` that just holds bit information: https://www.postgresql.org/docs/current/datatype-bit.html Very different.

Reply:
I thought about this some more, and I can't really see how to introduce a bit string type as an example without getting into some very specific use case.

I've rewritten using the `@db.Date` example I mentioned above - that's visually distinct and might be good enough, what do you think?